### PR TITLE
Do not cast .transform() output back to input dtype (closes #10972)

### LINKED ISSUE #10972: BUG: output of a transform is cast to dtype of input
xref #11444, #13046 for addtl tests

``` python
In [27]: df = pd.DataFrame({'a': np.random.randint(0, 5, 365), 'b': pd.date_range('2015-01-01', periods=365, freq='D')})

In [28]: df.head()
Out[28]:
   a          b
0  3 2015-01-01
1  3 2015-01-02
2  4 2015-01-03
3  2 2015-01-04
4  4 2015-01-05

In [29]: df.groupby('a').b.transform(lambda x: x.dt.dayofweek - x.dt.dayofweek.mean()).head()
Out[29]:
0   1970-01-01 00:00:00.000000000
1   1970-01-01 00:00:00.000000001
2   1970-01-01 00:00:00.000000001
3   1970-01-01 00:00:00.000000002
4   1969-12-31 23:59:59.999999997
Name: b, dtype: datetime64[ns]
```

I expected a float. No idea how difficult this will be so I marked it for 0.18. I won't have time to get to it any earlier, but if someone else wants to...

### PR DESCRIPTION
**Pull request by nbonnotte**
- [x] closes #10972
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

Maybe need more tests, and check performance?

### CODE REVIEW

**Review by jreback — COMMENTED** [review]
this needs perf  (just run the asv for groupby) and see

**Comment by nbonnotte** [comment]
Well, each time I run asv, I get different results, but after closing everything and leaving my computer alone, here is what I get:
```
   before     after       ratio
  [3853fe6d] [2bc47491]
+   19.84ms    22.47ms      1.13  groupby.GroupBySuite.time_diff('float', 100)
+   12.45ms    13.92ms      1.12  groupby.FirstLast.time_groupby_nth_none('float64')
-    1.97ms     1.78ms      0.90  groupby.GroupBySuite.time_size('int', 10000)
-   27.05ms    24.28ms      0.90  groupby.GroupBySuite.time_rank('float', 100)
-   19.79ms    17.21ms      0.87  groupby.FirstLast.time_groupby_nth_none('object')
-   18.12ms    15.64ms      0.86  groupby.FirstLast.time_groupby_nth_none('datetime')
-   10.96ms     9.37ms      0.86  groupby.FirstLast.time_groupby_last('datetime')
```

**Comment by jreback** [comment]
those are not the transform asv

**Comment by nbonnotte** [comment]
I run `asv continuous -f 1.1 -E virtualenv upstream/master HEAD -b ^groupby`, but it's the first time I do perf, so I may have missed something :)

Is there something else I should do for this PR?

**Comment by nbonnotte** [comment]
I think Appveyor is complaining about int being int32 on windows instead of int64. Is there a way to cast them to int64 instead?

File: pandas/tseries/tests/test_resample.py [hunk 1]
```diff
@@ -1805,7 +1805,7 @@ def test_resample_median_bug_1688(self):
                                           datetime(2012, 1, 1, 0, 5, 0)],
                            dtype=dtype)
 
-            result = df.resample("T").apply(lambda x: x.mean())
+            result = df.resample("T").mean()
             exp = df.asfreq('T')
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: why r u changing tests like this?

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -5669,7 +5668,7 @@ def test_ops_general(self):
         labels = np.random.randint(0, 50, size=1000).astype(float)
 
         for op, targop in ops:
-            result = getattr(df.groupby(labels), op)().astype(float)
+            result = getattr(df.groupby(labels), op)()
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: don't change tests

your changes are not robust

**Review by jreback — CHANGES_REQUESTED** [review]
don't change any tests; you have to maintain the existing unless their is a really good reason

tests breaking generally means something is wrong in your code 

i suspect your changes are not propagating dtypes correctly 

add the tests from the issue

**Comment by nbonnotte** [comment]
@jreback Hum, this is more complex than I thought

The initial problem is: `.transform` cast output dtypes back to input dtype, so `.size` for instance produce timestamps if the initial data was timestamps

If I keep the output type as it is, then `Series.mean`, for instance, will produce `float`. So if the original data is `float32`,  performing `.groupby().apply(lambda g: g.mean())`  might thus result in `float64` output, but `.groupby().mean()` will stay in `float32`.

Am I missing something obvious? How to proceed from there?

**Comment by jreback** [comment]
you need to selectively cast

all lost every function should be cast except for things that produce ints (like size)

**Comment by nbonnotte** [comment]
So `.groupby().agg(lambda : float(len(x)))` should keep producing timestamps if input is timestamps?

(That's how I stumbled on this issue when working on my PR #12607)

**Comment by jreback** [comment]
@nbonnotte of course not, that's the point of this issue!

you simply need to infer based on the return values, not automatically cast it.

you *might* be able to simply stick it into a Series and it will work, BUT, you will still need to cast if its a numeric type (IOW a float) to the original type flavor.

**Comment by nbonnotte** [comment]
Ok. I'm going to take it slow, starting with datetime input

**Comment by jreback** [comment]
this was fixed by https://github.com/pandas-dev/pandas/commit/251826f0861159160bd1d51eafadb6e0b4161f77

**Comment by nbonnotte** [comment]
It looks so simple once you see the solution... Thanks @jreback for your feedback and your patience, sorry I couldn't get that through

**Pull request closed without merging**